### PR TITLE
[GStreamer][VideoCapture] fix getVideoSizeAndFormatFromCaps/getVideoResolutionFromCaps for video/x-raw mime-type caps without format

### PR DESCRIPTION
**Pull request by jameshilliard**
#### fdb0e95ecfc0a8c011ed589562dd4162551d0f5a
<pre>
[GStreamer][VideoCapture] fix getVideoSizeAndFormatFromCaps/getVideoResolutionFromCaps for video/x-raw mime-type caps without format
<a href="https://bugs.webkit.org/show_bug.cgi?id=242290">https://bugs.webkit.org/show_bug.cgi?id=242290</a>

Reviewed by Philippe Normand.

Fixes:
0:00:02.001539663  1169 0x5577d2b74810 DEBUG             video-info video-info.c:406:gst_video_info_from_caps: parsing caps video/x-raw
0:00:02.001842884  1169 0x5577d2b74810 ERROR             video-info video-info.c:546:gst_video_info_from_caps: no format given
0:00:02.002197896  1169 0x5577d2b74810 INFO     webkitvideocapturer GStreamerVideoCapturer.cpp:136:setSize: Setting size to 640x480

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getVideoSizeAndFormatFromCaps):
(WebCore::getVideoResolutionFromCaps):

Canonical link: <a href="https://commits.webkit.org/252112@main">https://commits.webkit.org/252112@main</a>
</pre>

### CODE REVIEW

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp [hunk 1]
```diff
@@ -126,7 +126,7 @@ bool getVideoSizeAndFormatFromCaps(const GstCaps* caps, WebCore::IntSize& size,
 
         size.setWidth(width);
         size.setHeight(height);
-    } else {
+    } else if (!gst_structure_has_name(structure, "video/x-raw") || gst_structure_has_name(structure, "format")) {
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: In which case did you see a structure named `format`?

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp [hunk 2]
```diff
@@ -138,6 +138,19 @@ bool getVideoSizeAndFormatFromCaps(const GstCaps* caps, WebCore::IntSize& size,
         pixelAspectRatioNumerator = GST_VIDEO_INFO_PAR_N(&info);
         pixelAspectRatioDenominator = GST_VIDEO_INFO_PAR_D(&info);
         stride = GST_VIDEO_INFO_PLANE_STRIDE(&info, 0);
+    } else {
+        format = GST_VIDEO_FORMAT_UNKNOWN;
+        stride = 0;
+        int width = 0, height = 0;
+        gst_structure_get_int(structure, "width", &width);
+        gst_structure_get_int(structure, "height", &height);
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:136]: Return values should be checked.

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp [hunk 3]
```diff
@@ -153,12 +166,12 @@ std::optional<FloatSize> getVideoResolutionFromCaps(const GstCaps* caps)
...
     int pixelAspectRatioNumerator = 1, pixelAspectRatioDenominator = 1;
 
+    GstStructure* structure = gst_caps_get_structure(caps, 0);
     if (areEncryptedCaps(caps)) {
-        GstStructure* structure = gst_caps_get_structure(caps, 0);
         gst_structure_get_int(structure, "width", &width);
         gst_structure_get_int(structure, "height", &height);
         gst_structure_get_fraction(structure, "pixel-aspect-ratio", &pixelAspectRatioNumerator, &pixelAspectRatioDenominator);
-    } else {
+    } else if (!gst_structure_has_name(structure, "video/x-raw") || gst_structure_has_name(structure, "format")) {
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: Ditto. Did you mean to use `gst_structure_has_field()` here?

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp [hunk 4]
```diff
@@ -168,6 +181,10 @@ std::optional<FloatSize> getVideoResolutionFromCaps(const GstCaps* caps)
         height = GST_VIDEO_INFO_HEIGHT(&info);
         pixelAspectRatioNumerator = GST_VIDEO_INFO_PAR_N(&info);
         pixelAspectRatioDenominator = GST_VIDEO_INFO_PAR_D(&info);
+    } else {
+        gst_structure_get_int(structure, "width", &width);
+        gst_structure_get_int(structure, "height", &height);
+        gst_structure_get_fraction(structure, "pixel-aspect-ratio", &pixelAspectRatioNumerator, &pixelAspectRatioDenominator);
```
Comment (philn) [inline Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:173]: Return values should be checked.

**Review by philn — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp (hunk 1 above)
Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: The failure comes from [format being missing from here](https://github.com/GStreamer/gstreamer/blob/1.20.2/subprojects/gst-plugins-base/gst-libs/gst/video/video-info.c#L410) so checking `gst_structure_has_name(structure, "format")` guards against `gst_video_info_from_caps` being called in cases where `gst_video_info_from_caps` will fail due to not having `format`.

Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: `gst_structure_get_string` fetches a field of the struct, not its name.

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp (hunk 3 above)
Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: Oh, probably.

Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: Changed to `gst_structure_has_field`.

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp (hunk 2 above)
Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:136]: Hmm, these fallbacks are based off of the `areEncryptedCaps` paths which also don't have checked returns:
https://github.com/WebKit/WebKit/blob/16de8a90c10880777b7e106e3e11b313cca02951/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L120-L121

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp (hunk 1 above)
Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp]: Yeah, just noticed that, changed to `gst_structure_has_field`.

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp (hunk 4 above)
Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:173]: They don't seem to be getting checked elsewhere, not sure if that's actually needed here?:
https://github.com/WebKit/WebKit/blob/16de8a90c10880777b7e106e3e11b313cca02951/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L158-L160

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp [hunk 5]
```diff
@@ -112,21 +112,8 @@ bool getVideoSizeAndFormatFromCaps(const GstCaps* caps, WebCore::IntSize& size,
...
-        if (!gst_structure_get_fraction(structure, "pixel-aspect-ratio", &pixelAspectRatioNumerator, &pixelAspectRatioDenominator)) {
-            pixelAspectRatioNumerator = 1;
-            pixelAspectRatioDenominator = 1;
-        }
-
-        size.setWidth(width);
-        size.setHeight(height);
-    } else {
+    GstStructure* structure = gst_caps_get_structure(caps, 0);
+    if (!areEncryptedCaps(caps) && (!gst_structure_has_name(structure, "video/x-raw") || gst_structure_has_field(structure, "format"))) {
```
Comment (jameshilliard) [inline Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:116]: Refactored this a bit to combine the encrypted caps codepaths with the no format codepaths.

**Review by philn — APPROVED** [review]

File: Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp (hunk 5 above)
Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:116]: Why is `gst_video_info_from_caps()` not called for `video/x-raw` caps? This looks incorrect...

Comment (philn) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:116]: Rephrasing... with this change, `gst_video_info_from_caps()` can actually be called for non raw caps, which triggers errors.

Comment (jameshilliard) [reply Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:116]: > Rephrasing... with this change, `gst_video_info_from_caps()` can actually be called for non raw caps, which triggers errors.

I didn't change that behavior, it was always called from the else block before this change:
https://github.com/WebKit/WebKit/blob/d2b5507f6b8e81ec0d5bd4de2d3d583e3fd3d199/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L129-L133

**Pull request merged**